### ISSUE TITLE
Facebook native ads call to action button does not respond to taps

### ISSUE THREAD

**Issue by stsandro** [googleads/googleads-mobile-ios-mediation#198]
I'm finishing my Facebook native ads mediation integration and I've noticed during testing that the call to action button does not respond to taps when a facebook native ad is shown.

My setup is similar to the [NativeAdvancedExample](https://github.com/googleads/googleads-mobile-ios-examples/tree/master/Swift/admob/NativeAdvancedExample). The main difference is that I've added `pod 'GoogleMobileAdsMediationFacebook'` to the Podfile and that I'm using an ad unit id that has mediation with facebook.

The problem seems to be connected to the following code:
```Swift
// In order for the SDK to process touch events properly, user interaction should be disabled.  
nativeAdView.callToActionView?.isUserInteractionEnabled = false
```
This is done when a native ad comes in and the `GADUnifiedNativeAdView` is populated with the native ad's properties. I've taken this line of code from the NativeAdvancedExample above. As soon as I change this line to:
```Swift
nativeAdView.callToActionView?.isUserInteractionEnabled = true
```
everything works as expected with AdMob ads and with Facebook ads. Both respond to touches on the call to action button.

This is similar to the following issue from the past #92 

I can create a sample project if this is needed.

**Comment by neobie** [comment]
Looking forwards to any possible solutions too.

**Comment by chlandacastro** [comment]
Hello Sandro!  I would like to know if you were able to solve this issue.   I tested it with Facebook Adapter 5.8.0.2  and 5.7.1.2 (from around Early March) and I was not able to replicate this problem.   I testing with our testing app and also the NativeAdvancedExample app.
If you still have this issue, please provide your sample project with the adapter version.

**Comment by stsandro** [comment]
@chlandacastro I've tested with AdMob SDK 7.58.0 and Facebook adapter 5.8.0.2. You can try it yourself. I've created a [fork of the NativeAdsExample](https://github.com/stsandro/googleads-mobile-ios-examples) and tested with an ad unit id that has facebook in the mediation setup. Here are the steps to replicate the issue:

* Clone the forked [googleads-mobile-ios-examples](https://github.com/stsandro/googleads-mobile-ios-examples)
* Setup and open the example at /Swift/admob/NativeAdvancedExample/
* Change the adUnitID in the ViewController to one that has facebook ads mediation
* Add test device id(s) in the testing section of facebook ads to receive ads
* Install the app on a real device and wait for a facebook ad to appear

You'll notice that the CTA button does not respond to taps.

**Comment by shubhendu1988** [comment]
Any updates for this issue, as similar issue I am also facing, Using Google Ads Manager sdk ('Google-Mobile-Ads-SDK', '7.57.0') and show Facebook ads via mediation. On receiving FB Ads CTA does not work where as in receiving google ads CTA works. As per google documentation UserInteration for UIBUtton needs to be false. Making it true makes FB ads CTA function but stops Google ads CTA. 

The workaround was to use UILabel/UIImageView. But here the problem is to show animation on tapping UILabel. 

Please update

**Comment by stsandro** [comment]
I'm doing the following as a workaround for this issue: 
- check the `nativeAd.responseInfo.adNetworkClassName`
- set the `isUserInteractionEnabled` either to `true` or `false` depending on the `adNetworkClassName`

```Swift
func configureWithNativeAd(_ nativeAd: GADUnifiedNativeAd) {
  // Required properties to display (headline, body, cta)
  (headlineView as! UILabel).text = nativeAd.headline
  (bodyView as! UILabel).text = nativeAd.body
    
  (callToActionView as? UIButton)?.setTitle(nativeAd.callToAction, for: .normal)
  
  // do some more configuration...

  // workaround for CTA 
  if  nativeAd.responseInfo.adNetworkClassName == "GADMAdapterFacebook" {
    callToActionView?.isUserInteractionEnabled = true
  } else {
    callToActionView?.isUserInteractionEnabled = false
  }
}
```
I don't really like this solution, but for me it was the only simple and fast way to get this working. I'm very open for a better solution or a fix in the mediation adapter or the GoogleMobileAdsSDK

**Comment by ericleich** [comment]
We've synced with Facebook on this issue, and Facebook is planning a fix in an upcoming SDK update to effectively do @stsandro 's workaround. @shubhendu1988 's workaround is the other short term solution.

The problem ultimately stems from Google and Facebook having different expectations on userInteractionEnabled. The way Google's SDK adds touch gestures, it wants isUserInteractionEnabled = false. Facebook currently wants isUserInteractionEnabled = true for UIButtons. In general, it's ok for expectations to be different across networks.  Each asset view (e.g. call to action view) is available to an adapter to forward to its corresponding SDK, who can change this property however they'd like when it's their ad being shown. If your app reuses the native ad view and a new ad from a different network goes in, the next network gets a reference to the same set of views and has a chance to update userInteractionEnabled as they see fit. The Google Mobile Ads SDK does set userInteractionEnabled back to false (and logs a warning) when a Google ad is being rendered, but we have to rely on the partner SDKs to handle the view interaction as they see fit when it's their ad, as it's their SDK handling the click detection.

In this case, Facebook SDK is getting the reference to the callToActionView already. It just isn't currently changing isUserInteractionEnabled = true.

**Comment by stsandro** [comment]
I've tested with the most recent versions of Google Mobile Ads SDK (7.62.0) and Facebook Audience Network SDK (5.10.0) and as far as I can tell this issue is resolved. I've tested with my app and the the [Native Ad Example](https://github.com/googleads/googleads-mobile-ios-examples/tree/master/Swift/admob/NativeAdvancedExample) and everything seems to work. So 👍to Google and Facebook for resolving this issue.
